liblepton/scheme/conf/upgrade.scm

```diff
@@ -0,0 +1,233 @@
+
+  #:use-module  ( ice-9 format )
+  #:use-module  ( geda  config )
+  #:use-module  ( conf  keys   )
+
+  #:export      ( convert )
+)
+
+
+
+; public:
+;
+; convert(): upgrade gEDA configuration
+;
+; Read legacy gEDA configuration from geda.conf (in current
+; directory), geda-user.conf or geda-system.conf file,
+; convert it (using new names) and produce the corresponding
+; lepton*.conf configuration file.
+; We get conversion information (list of keys, old and new names)
+; from the [keys-list] defined in the (conf keys) module.
+;
+; [what]:                'local, 'user or 'system - config to convert
+; [report-absent-keys]:  print messages about missing keys in old cfg file
+; [overwrite]:           if #t, overwrite existing new cfg file
+; {thr}:                 'ctx 'infile 'load 'outfile 'save
+; {ret}:                 new cfg file full name or #f on failure
+;
+; This function might throw the following exceptions:
+; - 'ctx:      cannot get configuration context
+; - 'infile:   legacy config file does not exist
+; - 'load:     cannot read legacy config file
+; - 'outfile:  output config file already exists
+; - 'save:     cannot write output config file
+;
+( define* ( convert what #:key (report-absent-keys #f) (overwrite #f) )
+( let
+  (
+  ( cfg    #f )
+  ( fname  #f )
+  ( keys  '() )
+  ( data  '() )
+  )
+
+  ;
+  ; read legacy config file (geda*.conf):
+  ;
+
+  ( config-set-legacy-mode! #t ) ; use geda*.conf files
+
+  ( set! cfg ( get-cfg-ctx what ) )
+  ( unless cfg
+    ( throw 'ctx "Cannot get config context (input)" )
+  )
+
+  ( set! fname (config-filename cfg) )
+  ( unless ( and fname (access? fname F_OK) )
+    ( throw 'infile "Input config file does not exist:" fname )
+  )
+
+  ( format #t "ii: INPUT: ~a~%" fname )
+  ( format #t "~%" )
+
+  ( catch #t
+    ( lambda()
+      ( config-load! cfg )
+      ( set! keys (conf-keys-list) )
+      ( set! data
+        ( read-old cfg keys #:report-absent-keys report-absent-keys )
+      )
+    )
+    ( lambda( ex . args )
+      ( throw 'load "Cannot read config:" ex args )
+    )
+  )
+
+
+  ;
+  ; write new config file (lepton*.conf):
+  ;
+
+  ( format #t "~%" )
+
+  ( config-set-legacy-mode! #f ) ; use lepton*.conf files
+
+  ( set! cfg ( get-cfg-ctx what ) )
+  ( unless cfg
+    ( throw 'ctx "Cannot get config context (output)" )
+  )
+
+  ( set! fname (config-filename cfg) )
+  ( when ( and fname (access? fname F_OK) (not overwrite) )
+    ( throw 'outfile "Output config file already exists:" fname )
+  )
+
+  ( format #t "ii: OUTPUT: ~a~%" fname )
+
+  ( catch #t
+    ( lambda()
+      ( write-new cfg data )
+      ( config-save! cfg )
+    )
+    ( lambda( ex . args )
+      ( throw 'save "Cannot write config:" ex args )
+    )
+  )
+
+  ; return:
+  fname
+
+) ; let
+) ; convert()
+
+
+
+; get configuration context
+;
+; [what]: symbol: 'local, 'user or 'system
+; {ret}:  requested configuration context or #f on failure
+;
+( define ( get-cfg-ctx what )
```

Comment:
Here and below, and in other file code, defining of names after the code they are used in can lead to collisions that would be hard to find, if the names are already defined in some other imported code. I've already stumbled upon such situations in the past. So please avoid this.

Reply:
Thank you for the advice.
Code rearranged.